File: examples/fetch-forecasts.py
#!/usr/bin/env python3
"""
Example: Systematic fetching of PACE forecasting data

This script demonstrates how to:
1. Check for new forecast updates
2. Fetch latest forecasts and historical data
3. Fetch specific archived forecasts
4. Validate data quality
5. Store data for operational use

Usage:
    python examples/fetch-forecasts.py
    python examples/fetch-forecasts.py --period 2025-06
    python examples/fetch-forecasts.py --validate-only
"""

import pandas as pd
import requests
from datetime import datetime
import argparse
import json
from pathlib import Path

BASE_URL = "https://conflictlab.github.io/data/forecasts"


class PACEDataFetcher:
    """Client for fetching PACE forecasting data."""

    def __init__(self, base_url=BASE_URL, cache_dir="data/pace_cache"):
        self.base_url = base_url
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)

    def get_metadata(self, period="latest"):
        """Fetch forecast metadata."""
        url = f"{self.base_url}/{period}/metadata.json"
        response = requests.get(url)
        response.raise_for_status()
        return response.json()

    def check_for_update(self, last_fetch_date=None):
        """
        Check if new forecasts are available.

        Args:
            last_fetch_date: datetime or None - Last time forecasts were fetched

        Returns:
            (bool, datetime) - (has_update, run_date)
        """
        metadata = self.get_metadata("latest")
        run_date = datetime.fromisoformat(metadata['run_date'])

        if last_fetch_date is None:
            return True, run_date

        return run_date > last_fetch_date, run_date

    def fetch_forecasts(self, period="latest", horizon=12, include_intervals=True):
        """
        Fetch forecast data for a specific period.

        Args:
            period: 'latest' or 'YYYY-MM' (e.g., '2025-06')
            horizon: 6 or 12 months ahead
            include_intervals: Include min/max confidence intervals

        Returns:
            dict with keys: metadata, mean, min, max, historical
        """
        base = f"{self.base_url}/{period}"

        print(f"Fetching forecasts from {base}...")

        # Fetch metadata
        metadata = self.get_metadata(period)

        # Fetch forecasts
        mean = pd.read_csv(f"{base}/forecasts_h{horizon}.csv")

        result = {
            'metadata': metadata,
            'mean': mean,
            'period': period,
            'horizon': horizon
        }

        if include_intervals:
            result['min'] = pd.read_csv(f"{base}/forecasts_h{horizon}_min.csv")
            result['max'] = pd.read_csv(f"{base}/forecasts_h{horizon}_max.csv")

        # Fetch historical data
        result['historical'] = pd.read_csv(f"{base}/Hist.csv", parse_dates=[0])

        return result

    def fetch_latest(self, horizon=12, include_intervals=True):
        """Convenience method to fetch latest forecasts."""
        return self.fetch_forecasts("latest", horizon, include_intervals)

    def list_available_periods(self):
        """
        List all available forecast periods.

        Note: This requires fetching the archive directory listing,
        which isn't directly available via the static API.
        For now, we construct based on metadata.
        """
        # This would ideally come from a catalog.json endpoint
        # For now, return a note
        return "Use the website or GitHub repo to see available periods: " \
               "https://github.com/conflictlab/conflictlab.github.io/tree/main/public/data/forecasts/archive"

    def validate_data(self, data):
        """
        Validate fetched data for quality issues.

        Args:
            data: dict returned from fetch_forecasts()

        Returns:
            dict with validation results
        """
        results = {
            'valid': True,
            'warnings': [],
            'errors': []
        }

        metadata = data['metadata']
        mean = data['mean']
        historical = data['historical']

        # Check 1: Metadata consistency
        forecast_start = pd.to_datetime(metadata['forecast_start_date'])
        data_end = pd.to_datetime(metadata['historical_end_date'])

        expected_start = data_end + pd.DateOffset(months=1)
        if forecast_start != expected_start:
            results['errors'].append(
                f"Forecast start {forecast_start} != data_end + 1 month {expected_start}"
            )
            results['valid'] = False

        # Check 2: No trailing all-zero months in historical data
        last_row_sum = historical.iloc[-1, 1:].sum()
        if last_row_sum == 0:
            results['errors'].append(
                f"Last historical month ({historical.iloc[-1, 0]}) has all zeros"
            )
            results['valid'] = False

        # Check 3: Reasonable forecast values for active countries
        # Get last 12 months historical average
        last_12 = historical.iloc[-12:, 1:]
        hist_avg = last_12.mean()

        # Get forecast average
        forecast_avg = mean.iloc[:, 1:].mean()

        # Check for countries with significant history but zero forecast
        for country in hist_avg.index:
            if hist_avg[country] > 100 and forecast_avg.get(country, 0) == 0:
                results['warnings'].append(
                    f"{country}: {hist_avg[country]:.0f} avg historical but 0 forecast"
                )

        # Check 4: Forecast periods match metadata
        expected_periods = data['horizon']
        actual_periods = len(mean)
        if expected_periods != actual_periods:
            results['errors'].append(
                f"Expected {expected_periods} forecast periods, got {actual_periods}"
            )
            results['valid'] = False

        return results

    def save_to_cache(self, data, filename=None):
        """Save fetched data to local cache."""
        if filename is None:
            period = data['metadata']['forecast_start_date']
            horizon = data['horizon']
            filename = f"pace_forecast_{period}_h{horizon}.json"

        filepath = self.cache_dir / filename

        # Convert DataFrames to dict for JSON serialization
        cache_data = {
            'metadata': data['metadata'],
            'mean': data['mean'].to_dict('records'),
            'historical': data['historical'].to_dict('records'),
            'cached_at': datetime.now().isoformat()
        }

        if 'min' in data:
            cache_data['min'] = data['min'].to_dict('records')
        if 'max' in data:
            cache_data['max'] = data['max'].to_dict('records')

        with open(filepath, 'w') as f:
            json.dump(cache_data, f, indent=2)

        print(f"Saved to {filepath}")
        return filepath


def main():
    parser = argparse.ArgumentParser(
        description='Fetch PACE conflict forecasting data'
    )
    parser.add_argument(
        '--period',
        default='latest',
        help='Forecast period: "latest" or YYYY-MM (e.g., 2025-06)'
    )
    parser.add_argument(
        '--horizon',
        type=int,
        default=12,
        choices=[6, 12],
        help='Forecast horizon in months'
    )
    parser.add_argument(
        '--no-intervals',
        action='store_true',
        help='Skip min/max confidence intervals (faster)'
    )
    parser.add_argument(
        '--validate-only',
        action='store_true',
        help='Only validate latest data, don\'t save'
    )
    parser.add_argument(
        '--save',
        action='store_true',
        help='Save fetched data to cache'
    )

    args = parser.parse_args()

    # Initialize fetcher
    fetcher = PACEDataFetcher()

    # Check for updates
    print("Checking for updates...")
    has_update, run_date = fetcher.check_for_update()

    if has_update:
        print(f"✅ New forecasts available from {run_date}")
    else:
        print("ℹ️  No new forecasts since last check")

    # Fetch data
    print(f"\nFetching {args.period} forecasts (h={args.horizon})...")
    data = fetcher.fetch_forecasts(
        period=args.period,
        horizon=args.horizon,
        include_intervals=not args.no_intervals
    )

    # Print summary
    print(f"\n{'='*60}")
    print("FORECAST SUMMARY")
    print(f"{'='*60}")
    print(f"Period: {data['metadata']['forecast_start_date']} to {data['metadata'][f'h{args.horizon}_end_date']}")
    print(f"Data through: {data['metadata']['historical_end_date']}")
    print(f"Countries: {len(data['mean'].columns) - 1}")
    print(f"Historical months: {len(data['historical'])}")
    print(f"Forecast months: {len(data['mean'])}")

    # Validate
    print(f"\n{'='*60}")
    print("DATA VALIDATION")
    print(f"{'='*60}")
    validation = fetcher.validate_data(data)

    if validation['valid']:
        print("✅ All validation checks passed")
    else:
        print("❌ Validation failed:")
        for error in validation['errors']:
            print(f"  - {error}")

    if validation['warnings']:
        print("\n⚠️  Warnings:")
        for warning in validation['warnings'][:5]:
            print(f"  - {warning}")
        if len(validation['warnings']) > 5:
            print(f"  ... and {len(validation['warnings']) - 5} more")

    # Save if requested
    if args.save and not args.validate_only:
        fetcher.save_to_cache(data)

    # Show sample data
    print(f"\n{'='*60}")
    print("SAMPLE: First 3 Countries, First 3 Months")
    print(f"{'='*60}")
    sample = data['mean'].iloc[:3, :4]
    print(sample)

    if 'min' in data and 'max' in data:
        print("\nConfidence Intervals (90%):")
        countries = data['mean'].columns[1:4]
        for country in countries:
            mean_val = data['mean'][country].iloc[0]
            min_val = data['min'][country].iloc[0]
            max_val = data['max'][country].iloc[0]
            print(f"  {country}: {mean_val:.1f} ({min_val:.1f} - {max_val:.1f})")


if __name__ == '__main__':
    main()

File: examples/fetch-forecasts.R
#!/usr/bin/env Rscript
#' Systematic fetching of PACE forecasting data
#'
#' This script demonstrates how to:
#' 1. Check for new forecast updates
#' 2. Fetch latest forecasts and historical data
#' 3. Validate data quality
#' 4. Store data for operational use
#'
#' Usage:
#'   Rscript examples/fetch-forecasts.R
#'   Rscript examples/fetch-forecasts.R --period 2025-06
#'   Rscript examples/fetch-forecasts.R --validate-only

library(httr)
library(jsonlite)
library(dplyr)
library(lubridate)

BASE_URL <- "https://conflictlab.github.io/data/forecasts"

#' Fetch forecast metadata
#'
#' @param period Forecast period: 'latest' or YYYY-MM
#' @return List with metadata
get_metadata <- function(period = "latest") {
  url <- sprintf("%s/%s/metadata.json", BASE_URL, period)
  response <- GET(url)
  stop_for_status(response)
  fromJSON(content(response, "text"))
}

#' Check if new forecasts are available
#'
#' @param last_fetch_date POSIXct or NULL - Last fetch timestamp
#' @return List with has_update (logical) and run_date (POSIXct)
check_for_update <- function(last_fetch_date = NULL) {
  metadata <- get_metadata("latest")
  run_date <- ymd_hms(metadata$run_date)

  has_update <- if (is.null(last_fetch_date)) {
    TRUE
  } else {
    run_date > last_fetch_date
  }

  list(has_update = has_update, run_date = run_date)
}

#' Fetch forecast data
#'
#' @param period 'latest' or 'YYYY-MM'
#' @param horizon 6 or 12 months ahead
#' @param include_intervals Include min/max confidence intervals
#' @return List with metadata, mean, min, max, historical
fetch_forecasts <- function(period = "latest",
                            horizon = 12,
                            include_intervals = TRUE) {
  base <- sprintf("%s/%s", BASE_URL, period)

  cat(sprintf("Fetching forecasts from %s...\n", base))

  # Fetch metadata
  metadata <- get_metadata(period)

  # Fetch forecasts
  mean <- read.csv(sprintf("%s/forecasts_h%d.csv", base, horizon))

  result <- list(
    metadata = metadata,
    mean = mean,
    period = period,
    horizon = horizon
  )

  if (include_intervals) {
    result$min <- read.csv(sprintf("%s/forecasts_h%d_min.csv", base, horizon))
    result$max <- read.csv(sprintf("%s/forecasts_h%d_max.csv", base, horizon))
  }

  # Fetch historical data
  result$historical <- read.csv(sprintf("%s/Hist.csv", base))
  result$historical$date <- as.Date(result$historical$date)

  result
}

#' Validate fetched data for quality issues
#'
#' @param data List returned from fetch_forecasts()
#' @return List with valid (logical), warnings, errors
validate_data <- function(data) {
  results <- list(
    valid = TRUE,
    warnings = c(),
    errors = c()
  )

  metadata <- data$metadata
  mean_df <- data$mean
  hist_df <- data$historical

  # Check 1: Metadata consistency
  forecast_start <- as.Date(paste0(metadata$forecast_start_date, "-01"))
  data_end <- as.Date(paste0(metadata$historical_end_date, "-01"))

  expected_start <- data_end %m+% months(1)

  if (forecast_start != expected_start) {
    msg <- sprintf(
      "Forecast start %s != data_end + 1 month %s",
      forecast_start, expected_start
    )
    results$errors <- c(results$errors, msg)
    results$valid <- FALSE
  }

  # Check 2: No trailing all-zero months
  last_row_sum <- sum(hist_df[nrow(hist_df), -1], na.rm = TRUE)

  if (last_row_sum == 0) {
    msg <- sprintf(
      "Last historical month (%s) has all zeros",
      hist_df$date[nrow(hist_df)]
    )
    results$errors <- c(results$errors, msg)
    results$valid <- FALSE
  }

  # Check 3: Countries with significant history but zero forecast
  last_12 <- tail(hist_df, 12)
  hist_avg <- colMeans(last_12[, -1], na.rm = TRUE)
  forecast_avg <- colMeans(mean_df[, -1], na.rm = TRUE)

  for (country in names(hist_avg)) {
    if (!is.na(hist_avg[country]) && hist_avg[country] > 100) {
      if (is.na(forecast_avg[country]) || forecast_avg[country] == 0) {
        msg <- sprintf(
          "%s: %.0f avg historical but 0 forecast",
          country, hist_avg[country]
        )
        results$warnings <- c(results$warnings, msg)
      }
    }
  }

  # Check 4: Forecast periods match metadata
  expected_periods <- data$horizon
  actual_periods <- nrow(mean_df)

  if (expected_periods != actual_periods) {
    msg <- sprintf(
      "Expected %d forecast periods, got %d",
      expected_periods, actual_periods
    )
    results$errors <- c(results$errors, msg)
    results$valid <- FALSE
  }

  results
}

#' Print forecast summary
#'
#' @param data List returned from fetch_forecasts()
print_summary <- function(data) {
  cat(rep("=", 60), "\n", sep = "")
  cat("FORECAST SUMMARY\n")
  cat(rep("=", 60), "\n", sep = "")

  metadata <- data$metadata
  h <- data$horizon

  cat(sprintf("Period: %s to %s\n",
              metadata$forecast_start_date,
              metadata[[sprintf("h%d_end_date", h)]]))
  cat(sprintf("Data through: %s\n", metadata$historical_end_date))
  cat(sprintf("Countries: %d\n", ncol(data$mean) - 1))
  cat(sprintf("Historical months: %d\n", nrow(data$historical)))
  cat(sprintf("Forecast months: %d\n", nrow(data$mean)))
}

#' Print validation results
#'
#' @param validation List returned from validate_data()
print_validation <- function(validation) {
  cat("\n", rep("=", 60), "\n", sep = "")
  cat("DATA VALIDATION\n")
  cat(rep("=", 60), "\n", sep = "")

  if (validation$valid) {
    cat("✅ All validation checks passed\n")
  } else {
    cat("❌ Validation failed:\n")
    for (error in validation$errors) {
      cat(sprintf("  - %s\n", error))
    }
  }

  if (length(validation$warnings) > 0) {
    cat("\n⚠️  Warnings:\n")
    for (warning in head(validation$warnings, 5)) {
      cat(sprintf("  - %s\n", warning))
    }
    if (length(validation$warnings) > 5) {
      cat(sprintf("  ... and %d more\n", length(validation$warnings) - 5))
    }
  }
}

#' Print sample data
#'
#' @param data List returned from fetch_forecasts()
print_sample <- function(data) {
  cat("\n", rep("=", 60), "\n", sep = "")
  cat("SAMPLE: First 3 Countries, First 3 Months\n")
  cat(rep("=", 60), "\n", sep = "")

  sample <- data$mean[1:3, 1:4]
  print(sample)

  if (!is.null(data$min) && !is.null(data$max)) {
    cat("\nConfidence Intervals (90%):\n")
    countries <- names(data$mean)[2:4]

    for (country in countries) {
      mean_val <- data$mean[[country]][1]
      min_val <- data$min[[country]][1]
      max_val <- data$max[[country]][1]

      cat(sprintf("  %s: %.1f (%.1f - %.1f)\n",
                  country, mean_val, min_val, max_val))
    }
  }
}

# Main execution
main <- function() {
  # Parse command line arguments
  args <- commandArgs(trailingOnly = TRUE)

  period <- "latest"
  horizon <- 12
  validate_only <- FALSE

  if (length(args) > 0) {
    for (i in seq_along(args)) {
      if (args[i] == "--period" && i < length(args)) {
        period <- args[i + 1]
      } else if (args[i] == "--horizon" && i < length(args)) {
        horizon <- as.integer(args[i + 1])
      } else if (args[i] == "--validate-only") {
        validate_only <- TRUE
      }
    }
  }

  # Check for updates
  cat("Checking for updates...\n")
  update_check <- check_for_update()

  if (update_check$has_update) {
    cat(sprintf("✅ New forecasts available from %s\n", update_check$run_date))
  } else {
    cat("ℹ️  No new forecasts since last check\n")
  }

  # Fetch data
  cat(sprintf("\nFetching %s forecasts (h=%d)...\n", period, horizon))
  data <- fetch_forecasts(period = period, horizon = horizon)

  # Print summary
  print_summary(data)

  # Validate
  validation <- validate_data(data)
  print_validation(validation)

  # Print sample
  if (!validate_only) {
    print_sample(data)
  }

  invisible(data)
}

# Run if called as script
if (sys.nframe() == 0L) {
  main()
}
